server/server/src/test/kotlin/org/jetbrains/bsp/bazel/server/bsp/utils/DelimitedMessageReaderTest.kt

```diff
@@ -0,0 +1,157 @@
+package org.jetbrains.bsp.bazel.server.bsp.utils
+
+import com.google.devtools.build.lib.buildeventstream.BuildEventStreamProtos
+import com.google.protobuf.InvalidProtocolBufferException
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.runTest
+import org.jetbrains.bsp.bazel.commons.Constants
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.fail
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.PipedInputStream
+import java.io.PipedOutputStream
+import kotlin.math.pow
+
+class DelimitedMessageReaderTest {
+  @Test
+  fun `test delimited messages are read`() =
+    runTest {
+      val buildStartedEvents =
+        buildList {
+          (0..3).forEach { i ->
+            BuildEventStreamProtos.BuildEvent
+              .newBuilder()
+              .apply {
+                started =
+                  startedBuilder
+                    .apply {
+                      uuid = i.toString()
+                      command = Constants.BAZEL_BUILD_COMMAND
+                    }.build()
+              }.build()
+              .also {
+                add(it)
+              }
+          }
+        }
+      val ba =
+        ByteArrayOutputStream().let { baos ->
+          buildStartedEvents.forEach { buildEvent ->
+            buildEvent.writeDelimitedTo(baos)
+          }
+          baos.toByteArray()
+        }
+
+      val reader = DelimitedMessageReader(ByteArrayInputStream(ba), BuildEventStreamProtos.BuildEvent.parser())
+
+      val result = listOf(reader.nextMessage(), reader.nextMessage(), reader.nextMessage(), reader.nextMessage())
+
+      Assertions.assertEquals(buildStartedEvents, result)
+      Assertions.assertNull(reader.nextMessage())
+    }
+
+  @Test
+  fun `test partial message handling`() =
+    runTest {
+      val message =
+        BuildEventStreamProtos.BuildEvent
+          .newBuilder()
+          .apply {
+            started =
+              startedBuilder
+                .apply {
+                  uuid = "uuid"
+                  command = Constants.BAZEL_BUILD_COMMAND
+                }.build()
+          }.build()
+
+      val ba =
+        ByteArrayOutputStream().let {
+          message.writeDelimitedTo(it)
+          it.toByteArray()
+        }
+
+      val partial = ba.sliceArray(0..2)
+
+      val reader =
+        DelimitedMessageReader(
+          ByteArrayInputStream(partial),
+          BuildEventStreamProtos.BuildEvent.parser(),
+        )
+
+      try {
+        reader.nextMessage()
+        fail("Exception expected when a partial message was sent")
+      } catch (_: InvalidProtocolBufferException) {
+        // expected
+      }
```

Comment:
```suggestion
      assertThrows(InvalidProtocolBufferException::class.java) { reader.nextMessage() }
```

Reply:
I originally tried that, but nextMessage suspends, and I am blanking on how to pass the context down into the unit

Reply:
Ahh, so nevermind then :) perhaps there's a method in kotest for suspending functions, but it was a nit anyway